### PR TITLE
Lookup template function

### PR DESCRIPTION
**Pull request by raffaelespazzoli**
I created a lookup template function to address #949 .
The lookup up function takes for string parameters:

1. the apiVersion, mandatory
2. the Kind, mandatory
3. the namespace, optional
4. the name, optional

and returns a `map[string]interface{}`

if namespace is missing the lookup function assumes we are looking up a cluster resource.
if name is missing the lookup function assumes we are looking up an list of resources.

for example the expression to lookup the value of a loadbalancer service would be:

`{{ (index (lookup "v1" "Service" "mynamespace" "myservice").status.loadBalancer.ingress 0).ip }}`
 
The way the lookup function works is the following:
1. at init time the restClient is initialized. I assume helm has already a way to initialized the rest client, so I'd gladly reuse that approach is someone can indicate to me how to do it.
2. the lookup function is added to the template functions and can therefore be used.

This is a work in progress and I being a new contributor request some assistance (and patience) while I comply with all the rules.

### CODE REVIEW

**Comment by raffaelespazzoli** [comment]
@thomastaylor312 if you could help with two things:

1. when I renamed my commits to comply with DCO, I inadvertently brought in also the last commit from the master branch. Is there a way to drop that commit from the PR?
2. circle ci is failing because, I think, I have introduced a new command line flag. This is due to the fact that I am not creating the kube rest client the right way. I assume there is a way to create the client that uses already existing flags. Can you show me how to do it?

**Comment by thomastaylor312** [comment]
@raffaelespazzoli You should be able to rebase your branch off of master and it should take care of it. As for the command line flag, let me dig into it some more. I haven't touched the engine package in a while and there might be an easier way

**Comment by thomastaylor312** [comment]
Ok, so after digging around in the [dynamic client docs](https://godoc.org/k8s.io/client-go/dynamic) I'm thinking the best way around this is to add a new top level Render method (probably something like `RenderWithClient` or something like that) that takes a [`Factory` interface](https://github.com/helm/helm/blob/33338eab17c1fe966f95889dde8f40696cb388a3/pkg/kube/factory.go#L28) which should already exist when it is called from the `action` package in Helm. You might will probably need to expand the Helm `Factory` interface to include the `ToDiscoveryClient` method.

Any way you do it, I would recommend against trying to regenerate the config separately like you are doing it currently

**Comment by raffaelespazzoli** [comment]
@thomastaylor312 I fixed the rebase issue, thanks for the suggestion.

I also took a look at how the kube client config is generated/managed.
It seems to me that the kube client config is embedded in a large config object (`action.Configuration`) that is created at the cobra command instantiation and passed down to each command. But there doesn't seem to be a globally visible config object reachable from every part of the code. 
The engine package can be called by many actions but has no access to each action's config.
I maybe wrong in my summary, but this is what I understood reading the code. 
So, I need some help finding a way to make the config object or the more specific kube client Factory visible in the engine package.

**Comment by thomastaylor312** [comment]
The `action.Configuration` object has the `RESTClientGetter` which should get you what you need. You could pass the getter or the configuration object down into the engine. Still trying to think of other options

**Comment by raffaelespazzoli** [comment]
@thomastaylor312 you say I can pass the `action.Configuration` or `RESTClientGetter` to the engine package. How would you do that? Keep in mind I am really new to the helm code base. with my limited understanding it seems to me that one of the issues there is that template processing will be called by multiple actions (I can think at least of `install` and `template`). I'd like to find a solution where we wound't have to modify and keep track of all the actions that can use the `engine`. Would it be better to create a global variable for the `RESTClientGetter`?

**Comment by raffaelespazzoli** [comment]
@thomastaylor312  re-based and added the fix about getting the rest client from the action configuration.

File: pkg/engine/lookup_func.go
Comment (thomastaylor312) [inline pkg/engine/lookup_func.go]: Nit: It may be helpful to have a comment here to explain why we are making sure it doesn't contain a `/` to help anyone down the line

File: pkg/engine/lookup_func.go
Comment (thomastaylor312) [inline pkg/engine/lookup_func.go]: What should be the behavior if `namespaced` is true, but the `namespace` is an empty string? Will there be any problems when it uses the non-namespaced client?

Comment (thomastaylor312) [reply pkg/engine/lookup_func.go]: Nit: Also, consider just passing the namespace into `getDynamicClientOnKind` and have it handle this logic. This logic seems to be a bit out of place here

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
Hey @raffaelespazzoli! Finally back from vacation and looking through this again. This looks really concise and straightforward. I have a handful of requested changes and a few nits (that I'll leave up to you to decide if you want to fix). I'll start pinging some of the other maintainers to take a look at this. Although I don't see any problems security-wise as I look at this now, I want to make sure that we've made sure we understand any security implications of adding this feature

File: pkg/action/install.go
Comment (raffaelespazzoli) [reply pkg/action/install.go]: I was trying to maintain the previous behavior as much as possible. I don't have a specific opinion either way, so let me know how you prefer to handle this. It's been a while but some of the unit tests run without the client initialized. I don't remember exactly in which of the ifs they were failing.

File: pkg/engine/lookup_func.go
Comment (raffaelespazzoli) [reply pkg/engine/lookup_func.go]: removed.

File: pkg/engine/lookup_func.go
Comment (raffaelespazzoli) [reply pkg/engine/lookup_func.go]: done

File: pkg/engine/lookup_func.go
Comment (raffaelespazzoli) [reply pkg/engine/lookup_func.go]: added comment

File: pkg/engine/lookup_func.go
Comment (raffaelespazzoli) [reply pkg/engine/lookup_func.go]: not sure what you would like to see here. Also why this error in particular and not every one else?

File: pkg/engine/lookup_func.go
Comment (raffaelespazzoli) [reply pkg/engine/lookup_func.go]: maybe we should add this to the doc, this is the intended behavior:

- namespaced true + non-empty namespace + non empty name -> kubectl get xxx <name> -n <namespace> -> single object
- namespaced true + non-empty namespace + empty name -> kubectl get xxx -n <namespace> -> list
- namespaced true + empty namespace (name is ignored) -> kubectl get xxx --all-namespaces -> list
- namespaced false (namespace is ignored) + non-empty name -> kubectl get xxx <name> -> single object
- namespaced false (namespace is ignored) + empty name -> kubectl get xxx -> list

Please help me make sure the algorithm honors the expected behavior.
This should also explain why we need this logic here and not in the `getDynamicClientOnKind`: we need to handle non-namespaced resources.

**Comment by jdolitsky** [comment]
How will these values be rendered on `helm template`, with the consideration that the user would not have access to a running cluster?

**Comment by etsauer** [comment]
@jdolitsky I'm working with @raffaelespazzoli a bit on this, doing some use case testing. Currently `helm template` would throw an error if there's no access to a cluster:

```
$ helm template ./charts/lookup-test/
lookup_func.go:99: [ERROR] unable to retrieve resource list for: v1 , error: Unauthorized
lookup_func.go:73: [ERROR] unable to get apiresource from unstructured: /v1, Kind=Service , error Unauthorized
```

I personally don't see any issue with that as its reasonable to expect a running cluster if you are going to use the lookup function.

**Comment by thomastaylor312** [comment]
@etsauer Helm template should run successfully without needing a connection to a Kubernetes, otherwise it would be a backwards incompatible change. If `--validate` is passed, it should then connect to k8s. I would suggest a filler value for when you can't connect

File: pkg/engine/lookup_func.go
Comment (thomastaylor312) [reply pkg/engine/lookup_func.go]: It should be everywhere, but specifically here I am guessing the built in k8s error doesn't provide the context needed to the user. That is why I left it a nit. Definitely not required, but extra context similar to what is in the log message could be useful

**Comment by raffaelespazzoli** [comment]
@jdolitsky @etsauer @thomastaylor312 in the current implementation there is some logic (see the pkg/action/install.go file) to decide whether to inject or no the lookup function. The lookup function is injected if the restclient is initialized. From my experience the restclient is not initialized only during unit test and always when helm is run from the CLI. regardless the logic is there. If you run a template with the lookup function, without having injected the lookup function in the template engine (i.e. the restclient is not initialized), you get an golang template error as it is to be expected. 
If you run a template with the lookup function and the client is initialized but it gets an error when establishing the connection, you get an error as shown by @etsauer.
It is kind of implied that in order to use this function you need a valid connection. I'm open discussing returning a default value instead of an error but I don't think it would be a good approach. This is not like the `Capability` family of variables that can be initialized if there is a connection or get a default if there isn't. The lookup function really can lookup any value of any type. It can return arrays, maps, string integer boo leans etc., so there wouldn't even be a one size fits all default type/value combination. And templates that get a type they didn't expect would fail anyway.

File: pkg/engine/lookup_func.go
Comment (thomastaylor312) [reply pkg/engine/lookup_func.go]: Yep, that makes sense. Thanks for the explanation. We should definitely add this to the documentation when this get merged

File: pkg/action/install.go
Comment (raffaelespazzoli) [reply pkg/action/install.go]: changed to return an error

File: pkg/engine/lookup_func.go
Comment (raffaelespazzoli) [reply pkg/engine/lookup_func.go]: added the wrap

**Comment by raffaelespazzoli** [comment]
@thomastaylor312 all you comment should be addressed

**Comment by raffaelespazzoli** [comment]
@thomastaylor312 ping...

**Comment by raffaelespazzoli** [comment]
@thomastaylor312 ping...

**Comment by thomastaylor312** [comment]
@raffaelespazzoli Sorry. All back from the holidays and finally digging myself out of the last of the backlog. Re-reviewing now and bringing it back up in the dev call this week

**Review by thomastaylor312 — COMMENTED** [review]
This is looking good. Holding off on final approval until we can talk about what the behavior should be with `helm template` in the dev call

**Comment by technosophos** [comment]
I am the one who rejected a similar proposal a few years ago. I am in favor of going forward with this PR. But I will call out one of the reasons I rejected a similar proposal in 2016:

One of the ideas behind `helm rollback` was that we could move a release back into the state it was in at that time. This required that all of the main dependencies were controlled by the chart (because if there were external dependencies, a rollback could not be reliable).

This feature will definitely put rollbacks at risk. But my feeling is that over time we now have a lot of external dependencies anyway. (CRDs are an example) So I think that while we should be _aware_ of this and educate users about this, we should go ahead with this PR.

**Review by thomastaylor312 — APPROVED** [review]
Also approving as is. After a good discussion in the dev call, we were in agreement that your logic is correct. If there are lookups in a chart, it should fail if used with `helm template` (without the `--validate` flag). So this LGTM. We will need a follow up doc PR in the docs repo to document all the things here

**Comment by thomastaylor312** [comment]
@raffaelespazzoli If you are feeling good with merging this, please remove the "WIP" and we'll go ahead and merge

**Comment by raffaelespazzoli** [comment]
@thomastaylor312 WIP removed, please merge.

**Comment by jpettit** [comment]
@raffaelespazzoli 

Just in case anyone else is wondering... the original proposed solution only works for `type: LoadBalancer`.  With a ClusterIP you'll get a nil record for `.status.loadBalancer.ingress`.

When using a `ClusterIP` it's a similar simple lookup:

`{{ (lookup "v1" "Service" "namespace" "service name").spec.clusterIP }}`

**Comment by huineng** [comment]
is this working in a post-install,post-upgrade method ?
i have a chart that creates a loadBalancer service , it has an ip
nevertheless when i want to create a configmap in a post hook to fill in that ip , i cannot get it to work

```
{{- if (lookup "v1" "Service" .Release.Namespace $service).status }}
  NATS__IP: {{ (index (lookup "v1" "Service" .Release.Namespace $service).status.loadBalancer.ingress 0).ip }}
  {{- else }}
  NATS__IP: 0.0.0.0
  {{- end }}
```

if i don't do the if i get 
```
at <$service>: nil pointer evaluating interface {}.loadBalancer
```

if i just do 
```
lookup "v1" "Service" .Release.Namespace $service
```

i get the value map[]  which means it's empty

so I4m wondering if the lookup is not immediately done instead if being done as part of the post hook

thanks

**Comment by BAGELreflex** [comment]
I wanted to do the same thing as huineng however it appears the `lookup` occurs prior to deploying any of the k8s resources defined in your template. This includes the post-install jobs too, the lookup / template resolution is occurring before any resources are created, even though it's a post-install job. I have instead just created a basic job which curl's the load balancer IP address from the local kubernetes API. This required me to create a ClusterRoleBinding and ClusterRole (I suspect RoleBinding and Role would have worked too) with permission to get information about the service.

**Pull request merged**
